urlc3ca8aa24b_davidsan_rivercityransom_tgz-pJ8-rivercityransom_dirtytests_GangsterTestMainJ8/src/rivercityransom/impl/GestionCombatImpl.java

```diff
@@ -0,0 +1,390 @@
+
+		// un certain nombre de gangsters aux noms poétiques
+		// String[] poetes = { "Guillaume Appolinaire", "Charles Baudelaire",
+		// "Paul Eluard", "Jacques Prévert", "Raymond Queneau",
+		// "Paul Verlaine", "Victor Hugo", "Alfred De Musset",
+		// "Arthur Rimbaud" };
+
+		String[] poetes = { "Bot0" };
+
+		gangster = new ArrayList<GangsterService>();
+		for (int i = 0; i < poetes.length; i++) { // 10 gangsters + slick
+			GangsterService g = new GangsterContract(new GangsterImpl());
+			// int positionGangsterX = largeurTerrain - 2 - i;
+			int positionGangsterX = RandomSingleton.getInstance().nextInt(
+					largeurTerrain);
+			int positionGangsterY = 0;
+			int positionGangsterZ = RandomSingleton.getInstance().nextInt(
+					profondeurTerrain);
+			ObjetService obj = new ObjetContract(new ObjetImpl());
+			if (i % 2 == 0) {
+				obj.init(Tresor.CHAINEDEVELO);
+			} else {
+				obj.init(Tresor.POUBELLEMETALLIQUE);
+			}
+			g.init(poetes[i], 16, 51, 16, 10, 50, positionGangsterX,
+					positionGangsterY, positionGangsterZ, obj);
+			gangster.add(g);
+			noms.add(poetes[i]);
+		}
+
+		// init hashes
+		estFrappeHash = new HashMap<String, Boolean>();
+		estGeleHash = new HashMap<String, Integer>();
+		for (String nom : noms) {
+			estFrappeHash.put(nom, false);
+			estGeleHash.put(nom, 0);
+		}
+	}
+
+	@Override
+	public void gerer(Commande commandRyan, Commande commandAlex) {
+		// generation d'une hash commande qui donne la commande associé à un
+		// personnage ou à un ennemi
+
+		HashMap<String, Commande> commandes = new HashMap<String, Commande>();
+		PersonnageService[] goodguys = { ryan(), alex() };
+		commandes.put("Ryan", commandRyan);
+		commandes.put("Alex", commandAlex);
+
+		List<GangsterService> badguys = new ArrayList<GangsterService>(
+				gangster());
+		badguys.add(slick());
+		RandomSingleton rs = RandomSingleton.getInstance();
+		Commande[] enumCommande = Commande.class.getEnumConstants();
+		for (GangsterService g : badguys) {
+			Commande c = enumCommande[rs.nextInt(enumCommande.length)];
+			commandes.put(g.nom(), c);
+		}
+
+		// mise à jour de la hash estFrappe
+		for (PersonnageService p : goodguys) {
+			if (p.estVaincu()) {
+				continue; // ignorons les morts
+			}
+			// trouver le gangster avec qui il est en collision
+			List<GangsterService> collidingGangster = collisionPerso(p);
+			estFrappeHash.put(p.nom(), false);
+			if (collidingGangster == null) {
+				continue; // pas de collision
+			} else {
+				for (GangsterService coll : collidingGangster) {
+					// il existe un gangster en collision
+					// vérifions qu'il n'est pas gelé
+					if (estGele(coll.nom())) {
+						continue;
+					}
+					// vérifions qu'il est en train de taper
+					if (commandes.get(coll.nom()) == Commande.FRAPPE) {
+						int degat = coll.force();
+						if (coll.estEquipe()) {
+							degat = degat * 2;
+						}
+						// mise à jour des points de vie
+						p.retrait(degat);
+						estFrappeHash.put(p.nom(), true);
+					}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 219:227, 250:258.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.